exporter/launch_failures.py

```diff
@@ -0,0 +1,82 @@
+from base import OSBase
+from osclient import session_adapter
+from prometheus_client import CollectorRegistry, generate_latest, Gauge
+from utils import node_details
+from datetime import datetime, timedelta
+from collections import Counter
+import logging
+
+logger = logging.getLogger(__name__)
+
+LABELS = [
+    'region',
+    'project_id'
+    'project_name']
+
+
+class LaunchFailures(OSBase):
+    """Class to report the statistics on OpenStack Nodes"""
+
+    def __init__(self, oscache, osclient):
+        super(LaunchFailures, self).__init__(oscache, osclient)
+        self.refresh_interval = oscache.refresh_interval
+
+    def build_cache_data(self):
+        """Return list of stats to cache."""
+        nova_api = session_adapter('compute')
+        keystone_api = session_adapter('identity')
+
+        nodes = node_details.get_nodes(detail=True)
+        servers = nova_api.get(
+            'servers/detail?all_tenants=True').json()['servers']
+
+        projects = keystone_api.get('v3/projects').json()['projects']
+        project_names = {p['id']: p['name'] for p in projects}
+
+        instance_failures = [
+            (s['project_id'], project_names[s['project_id']])
+            for s in servers
+            if self._valid_failure(s['status'] == 'ERROR', s['updated'])]
+
+        node_failures = [
+            (n.project_id, n.project_name)
+            for n in nodes
+            if self._valid_failure(n.last_error is not None, n.updated)
+        ]
+
+        counts = Counter(instance_failures + node_failures)
+
+        return [self._apply_labels(k, v) for k, v in counts.items()]
+
+    def _valid_failure(error, update_time):
+        update_time = datetime.strptime(
+            update_time, '%Y-%m-%dT%H:%M:%SZ')
+        time_since_scrape = (
+            datetime.utcnow() - timedelta(0, self.refresh_interval))
+
+        return error and (update_time > time_since_scrape)
+
+    def _apply_labels(project_info, count):
+        return dict(
+            project_id=project_info[0],
+            project_name=project_info[1],
+            stat_value=count
+        )
+
+    def get_cache_key(self):
+        return 'launch_failures'
+
+    def get_stats(self):
+        registry = CollectorRegistry()
+        labels = LABELS
+        launch_failures_cache = self.get_cache_data()
+        for launch_failure in launch_failures_cache:
+            stat_gauge = Gauge(
```

Comment:
You want this to be a `Counter` I think. It will be way simpler to just increment based on how many new server/node failures you found than keep track of the total # currently in play at a given time. Plus, I don't think we're tracking if a server/node comes out of an error state, so the Gauge value would never go down anyways (which is why you would use a gauge, if the value could go up or down and you're interested in the absolute value at a time, rather than a rate of change.)